File: src/bot.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const TelegramBot = require('node-telegram-bot-api');
const config = require('./lib/config');
const tmux = require('./lib/tmux');
const formatter = require('./lib/formatter');
const sessions = require('./lib/sessions');
const { logEvent, logError } = require('./lib/logger');

const cfg = config.loadConfig();
const bot = new TelegramBot(cfg.botToken, { polling: true });
const botStartTime = Math.floor(Date.now() / 1000);

bot.on('message', (msg) => {
  if (msg.date < botStartTime) return;

  const chatId = msg.chat.id;

  if (!cfg.chatId) {
    cfg.chatId = String(chatId);
    config.saveChatId(chatId);
    bot.sendMessage(
      chatId,
      '\u2705 Chat ID saved. You will now receive ' +
        'Claude Code notifications here.\n\n' +
        'Send /help to see available commands.'
    );
    return;
  }

  if (String(chatId) !== String(cfg.chatId)) {
    bot.sendMessage(
      chatId,
      '\u26D4 Unauthorized. This bot is configured ' + 'for a different chat.'
    );
    return;
  }

  if (msg.photo && msg.photo.length > 0) {
    handlePhoto(chatId, msg);
    return;
  }

  const text = msg.text;
  if (!text) return;

  const active = getActiveSession();
  if (text.startsWith('/')) {
    const cmd = text.split(/\s+/)[0];
    logEvent(active, `command: ${cmd}`);
  }

  if (text === '/stop') {
    handleStop(chatId);
  } else if (text === '/allow') {
    handleAllow(chatId);
  } else if (text === '/deny') {
    handleDeny(chatId);
  } else if (text === '/escape') {
    handleEscape(chatId);
  } else if (text === '/status') {
    handleStatus(chatId);
  } else if (text === '/screen') {
    handleScreen(chatId);
  } else if (text === '/plan') {
    handlePlan(chatId);
  } else if (text === '/sessions') {
    handleSessions(chatId);
  } else if (text.startsWith('/switch')) {
    handleSwitch(chatId, text);
  } else if (text.startsWith('/rename')) {
    handleRename(chatId, text);
  } else if (text === '/help') {
    handleHelp(chatId);
  } else if (/^\/\d+$/.test(text)) {
    handleSelectOption(chatId, parseInt(text.slice(1), 10));
  } else {
    handleInput(chatId, text);
  }
});

function getActiveSession() {
  return sessions.getActive();
}

function handleInput(chatId, text) {
  const active = getActiveSession();
  if (!active) {
    bot.sendMessage(
      chatId,
      '\u26A0\uFE0F No active sessions. Start one with: ' +
        '<code>claude-tg</code>',
      { parse_mode: 'HTML' }
    );
    return;
  }

  if (!tmux.sessionExists(active)) {
    sessions.unregister(active);
    bot.sendMessage(
      chatId,
      `\u26A0\uFE0F Session <code>${formatter.escapeHtml(active)}</code> is no longer running.`,
      { parse_mode: 'HTML' }
    );
    return;
  }

  try {
    tmux.sendKeys(active, text);
    logEvent(active, 'input sent to tmux');
    bot.sendMessage(
      chatId,
      `\uD83D\uDCE4 [${formatter.escapeHtml(active)}] Sent:\n<code>${formatter.escapeHtml(text)}</code>`,
      { parse_mode: 'HTML' }
    );
  } catch (err) {
    logError(active, `sendKeys failed: ${err.message}`);
    bot.sendMessage(chatId, `\u274C Failed to send: ${err.message}`);
  }
}

function handlePhoto(chatId, msg) {
  const active = getActiveSession();
  if (!active) {
    bot.sendMessage(
      chatId,
      '\u26A0\uFE0F No active sessions. Start one with: ' +
        '<code>claude-tg</code>',
      { parse_mode: 'HTML' }
    );
    return;
  }

  if (!tmux.sessionExists(active)) {
    sessions.unregister(active);
    bot.sendMessage(
      chatId,
      `\u26A0\uFE0F Session <code>${formatter.escapeHtml(active)}</code> is no longer running.`,
      { parse_mode: 'HTML' }
    );
    return;
  }

  const photo = msg.photo[msg.photo.length - 1];
  const caption = msg.caption || '';

  const uploadsDir = path.join(
    process.env.HOME,
    '.claude',
    'claude-tg',
    'uploads'
  );
  if (!fs.existsSync(uploadsDir)) {
    fs.mkdirSync(uploadsDir, { recursive: true });
  }

  bot
    .getFile(photo.file_id)
    .then((fileInfo) => {
      const ext = path.extname(fileInfo.file_path || '') || '.jpg';
      const filename = `telegram-${Date.now()}${ext}`;
      const savePath = path.join(uploadsDir, filename);

      return bot
        .downloadFile(photo.file_id, uploadsDir)
        .then((downloadedPath) => {
          if (downloadedPath !== savePath) {
            fs.renameSync(downloadedPath, savePath);
          }

          const input = caption
            ? `${caption} (see image: ${savePath})`
            : `Please look at this image: ${savePath}`;

          tmux.sendKeys(active, input);
          bot.sendMessage(
            chatId,
            `\uD83D\uDCF7 [${formatter.escapeHtml(active)}] Saved photo to <code>${formatter.escapeHtml(filename)}</code> and sent to Claude.`,
            { parse_mode: 'HTML' }
          );
        });
    })
    .catch((err) => {
      logError(active, `photo download failed: ${err.message}`);
      bot.sendMessage(
        chatId,
        `\u274C Failed to download photo: ${err.message}`
      );
    });
}

function handleStatus(chatId) {
  const info = sessions.list();
  const names = Object.keys(info.sessions);

  if (names.length === 0) {
    bot.sendMessage(chatId, '\uD83D\uDD34 No active sessions.');
    return;
  }

  const lines = [`\uD83D\uDFE2 <b>${names.length} active session(s):</b>`, ''];
  for (const name of names) {
    const alive = tmux.sessionExists(name);
    const icon = alive ? '\uD83D\uDFE2' : '\uD83D\uDD34';
    if (name === info.active) {
      lines.push(`${icon} <code>${formatter.escapeHtml(name)}</code> \u25C0 active`);
    } else {
      lines.push(`${icon} <code>/switch ${formatter.escapeHtml(name)}</code>`);
    }
  }

  bot.sendMessage(chatId, lines.join('\n'), {
    parse_mode: 'HTML',
  });
}

function handleSessions(chatId) {
  handleStatus(chatId);
}

function handleSwitch(chatId, text) {
  const parts = text.split(/\s+/);
  const target = parts[1];

  if (!target) {
    bot.sendMessage(chatId, 'Usage: <code>/switch claude-0214-1352</code>', {
      parse_mode: 'HTML',
    });
    return;
  }

  const info = sessions.list();
  if (!info.sessions[target]) {
    const available = Object.keys(info.sessions)
      .map((n) => `<code>${formatter.escapeHtml(n)}</code>`)
      .join(', ');
    bot.sendMessage(
      chatId,
      `\u26A0\uFE0F Session <code>${formatter.escapeHtml(target)}</code> not found.\n\nAvailable: ${available}`,
      { parse_mode: 'HTML' }
    );
    return;
  }

  sessions.setActive(target);
  bot.sendMessage(
    chatId,
    `\u2705 Switched to <code>${formatter.escapeHtml(target)}</code>`,
    { parse_mode: 'HTML' }
  );
}

function handleScreen(chatId) {
  const active = getActiveSession();
  if (!active) {
    bot.sendMessage(chatId, '\u26A0\uFE0F No active sessions.');
    return;
  }

  if (!tmux.sessionExists(active)) {
    bot.sendMessage(
      chatId,
      `\u26A0\uFE0F Session <code>${formatter.escapeHtml(active)}</code> is no longer running.`,
      { parse_mode: 'HTML' }
    );
    return;
  }

  try {
    const output = tmux.capturePane(active, 40);
    const trimmed = output.trim();
    if (!trimmed) {
      bot.sendMessage(
        chatId,
        `<i>[${formatter.escapeHtml(active)}] Screen is empty</i>`,
        { parse_mode: 'HTML' }
      );
      return;
    }
    const truncated =
      trimmed.length > 3800 ? `...\n${trimmed.slice(-3800)}` : trimmed;
    bot.sendMessage(
      chatId,
      `<b>[${formatter.escapeHtml(active)}]</b>\n<pre>${formatter.escapeHtml(truncated)}</pre>`,
      { parse_mode: 'HTML' }
    );
  } catch (err) {
    logError(active, `screen capture failed: ${err.message}`);
    bot.sendMessage(chatId, `\u274C Failed to capture screen: ${err.message}`);
  }
}

function handlePlan(chatId) {
  const active = getActiveSession();
  if (!active) {
    bot.sendMessage(chatId, '\u26A0\uFE0F No active sessions.');
    return;
  }

  if (!tmux.sessionExists(active)) {
    bot.sendMessage(chatId, '\u26A0\uFE0F Session not found.');
    return;
  }

  try {
    const output = tmux.capturePane(active, 50);
    const match = output.match(/~\/\.claude\/plans\/[^\s]+\.md/);
    if (!match) {
      bot.sendMessage(
        chatId,
        '\u26A0\uFE0F No plan file found in terminal. Is there a plan approval prompt visible?'
      );
      return;
    }

    const planPath = match[0].replace(/^~/, process.env.HOME);
    const content = fs.readFileSync(planPath, 'utf8').trim();
    if (!content) {
      bot.sendMessage(chatId, '<i>Plan file is empty.</i>', {
        parse_mode: 'HTML',
      });
      return;
    }

    logEvent(active, `viewed plan: ${match[0]}`);
    sendLongMessage(
      chatId,
      `\uD83D\uDCCB [${formatter.escapeHtml(active)}] Plan`,
      content
    );
  } catch (err) {
    logError(active, `plan read failed: ${err.message}`);
    bot.sendMessage(chatId, `\u274C Failed to read plan: ${err.message}`);
  }
}

function sendLongMessage(chatId, title, content) {
  const escaped = formatter.escapeHtml(content);
  const maxChunk = 3800;

  // Try single message first
  const single = `<b>${title}</b>\n<pre>${escaped}</pre>`;
  if (single.length <= 4096) {
    bot.sendMessage(chatId, single, { parse_mode: 'HTML' });
    return;
  }

  // Split on line boundaries
  const lines = escaped.split('\n');
  const chunks = [];
  let current = '';

  for (const line of lines) {
    if (current.length + line.length + 1 > maxChunk && current.length > 0) {
      chunks.push(current);
      current = line;
    } else {
      current += (current ? '\n' : '') + line;
    }
  }
  if (current) chunks.push(current);

  // Send first chunk with title
  bot.sendMessage(
    chatId,
    `<b>${title}</b>\n<pre>${chunks[0]}</pre>`,
    { parse_mode: 'HTML' }
  );

  // Send remaining chunks
  for (let i = 1; i < chunks.length; i++) {
    bot.sendMessage(chatId, `<pre>${chunks[i]}</pre>`, {
      parse_mode: 'HTML',
    });
  }
}

function handleRename(chatId, text) {
  const parts = text.split(/\s+/);
  const newName = parts[1];

  if (!newName) {
    bot.sendMessage(chatId, 'Usage: <code>/rename my-project</code>', {
      parse_mode: 'HTML',
    });
    return;
  }

  if (!/^[a-zA-Z0-9][a-zA-Z0-9-]*$/.test(newName)) {
    bot.sendMessage(
      chatId,
      '\u26A0\uFE0F Invalid name. Use only letters, ' + 'numbers, and hyphens.'
    );
    return;
  }

  const active = getActiveSession();
  if (!active) {
    bot.sendMessage(chatId, '\u26A0\uFE0F No active sessions.');
    return;
  }

  const info = sessions.list();
  if (info.sessions[newName]) {
    bot.sendMessage(
      chatId,
      `\u26A0\uFE0F Name <code>${formatter.escapeHtml(newName)}</code> is already in use.`,
      { parse_mode: 'HTML' }
    );
    return;
  }

  try {
    tmux.renameSession(active, newName);
    sessions.rename(active, newName);
    bot.sendMessage(
      chatId,
      `\u2705 Renamed <code>${formatter.escapeHtml(active)}</code> \u2192 <code>${formatter.escapeHtml(newName)}</code>`,
      { parse_mode: 'HTML' }
    );
  } catch (err) {
    logError(active, `rename failed: ${err.message}`);
    bot.sendMessage(chatId, `\u274C Failed to rename: ${err.message}`);
  }
}

function handleStop(chatId) {
  const active = getActiveSession();
  if (!active) {
    bot.sendMessage(chatId, '\u26A0\uFE0F No active sessions.');
    return;
  }

  if (!tmux.sessionExists(active)) {
    bot.sendMessage(chatId, '\u26A0\uFE0F Session not found.');
    return;
  }

  try {
    tmux.sendInterrupt(active);
    bot.sendMessage(
      chatId,
      `\u23F9 [${formatter.escapeHtml(active)}] Sent Ctrl+C`,
      { parse_mode: 'HTML' }
    );
  } catch (err) {
    bot.sendMessage(chatId, `\u274C Failed: ${err.message}`);
  }
}

function handleAllow(chatId) {
  const active = getActiveSession();
  if (!active) {
    bot.sendMessage(chatId, '\u26A0\uFE0F No active sessions.');
    return;
  }

  if (!tmux.sessionExists(active)) {
    bot.sendMessage(chatId, '\u26A0\uFE0F Session not found.');
    return;
  }

  try {
    tmux.sendEnter(active);
    bot.sendMessage(
      chatId,
      `\u2705 [${formatter.escapeHtml(active)}] Approved permission`,
      { parse_mode: 'HTML' }
    );
  } catch (err) {
    bot.sendMessage(chatId, `\u274C Failed: ${err.message}`);
  }
}

function handleDeny(chatId) {
  const active = getActiveSession();
  if (!active) {
    bot.sendMessage(chatId, '\u26A0\uFE0F No active sessions.');
    return;
  }

  if (!tmux.sessionExists(active)) {
    bot.sendMessage(chatId, '\u26A0\uFE0F Session not found.');
    return;
  }

  try {
    tmux.sendEscape(active);
    bot.sendMessage(
      chatId,
      `\u274C [${formatter.escapeHtml(active)}] Denied permission`,
      { parse_mode: 'HTML' }
    );
  } catch (err) {
    bot.sendMessage(chatId, `\u274C Failed: ${err.message}`);
  }
}

function handleSelectOption(chatId, optionNum) {
  const active = getActiveSession();
  if (!active) {
    bot.sendMessage(chatId, '\u26A0\uFE0F No active sessions.');
    return;
  }

  if (!tmux.sessionExists(active)) {
    bot.sendMessage(chatId, '\u26A0\uFE0F Session not found.');
    return;
  }

  try {
    // Navigate to top (10 Ups is enough for any menu), then down to target
    tmux.sendArrowUp(active, 10);
    if (optionNum > 1) {
      tmux.sendArrowDown(active, optionNum - 1);
    }
    tmux.sendEnter(active);
    logEvent(active, `selected option ${optionNum}`);
    bot.sendMessage(
      chatId,
      `\u2705 [${formatter.escapeHtml(active)}] Selected option ${optionNum}`,
      { parse_mode: 'HTML' }
    );
  } catch (err) {
    logError(active, `select option failed: ${err.message}`);
    bot.sendMessage(chatId, `\u274C Failed: ${err.message}`);
  }
}

function handleEscape(chatId) {
  const active = getActiveSession();
  if (!active) {
    bot.sendMessage(chatId, '\u26A0\uFE0F No active sessions.');
    return;
  }

  if (!tmux.sessionExists(active)) {
    bot.sendMessage(chatId, '\u26A0\uFE0F Session not found.');
    return;
  }

  try {
    tmux.sendEscape(active);
    bot.sendMessage(
      chatId,
      `\u23F9 [${formatter.escapeHtml(active)}] Sent Escape`,
      { parse_mode: 'HTML' }
    );
  } catch (err) {
    bot.sendMessage(chatId, `\u274C Failed: ${err.message}`);
  }
}

function handleHelp(chatId) {
  bot.sendMessage(
    chatId,
    [
      '<b>claude-tg commands:</b>',
      '',
      '/allow - Approve a permission prompt',
      '/deny - Deny a permission prompt',
      '/1, /2, ... - Select a numbered option',
      '/stop - Send Ctrl+C to interrupt Claude',
      '/escape - Send Escape key',
      '/status - Show all active sessions',
      '/sessions - Same as /status',
      '/switch &lt;name&gt; - Switch active session',
      '/rename &lt;name&gt; - Rename the active session',
      '/screen - Show recent terminal output',
      '/plan - View the current plan',
      '/help - Show this message',
      '',
      'Any other text is sent as input to the ' + 'active Claude session.',
    ].join('\n'),
    { parse_mode: 'HTML' }
  );
}

function shutdown() {
  logEvent(null, 'bot shutting down');
  console.log('Shutting down...');
  bot.stopPolling();
  process.exit(0);
}

process.on('SIGINT', shutdown);
process.on('SIGTERM', shutdown);

bot.on('polling_error', (err) => {
  if (
    err.code === 'ETELEGRAM' &&
    err.response &&
    err.response.statusCode === 409
  ) {
    logError(null, 'duplicate bot instance detected, exiting');
    console.error(
      'ERROR: Another bot instance is running ' +
        'with the same token. Exiting.'
    );
    process.exit(1);
  }
  logError(null, `polling error: ${err.message}`);
  console.error('Polling error:', err.message);
});

logEvent(null, 'bot started');
console.log('claude-tg bot is running. Waiting for messages...');

File: internal/bot/bot.go
package bot

import (
	"fmt"
	"log"
	"os"
	"os/signal"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"syscall"
	"time"

	"claude-tg/internal/config"
	"claude-tg/internal/formatter"
	"claude-tg/internal/sessions"
	"claude-tg/internal/telegram"
	"claude-tg/internal/tmux"
)

var nameRegex = regexp.MustCompile(`^[a-zA-Z0-9][a-zA-Z0-9-]*$`)
var optionRegex = regexp.MustCompile(`^/(\d+)$`)
var planPathRegex = regexp.MustCompile(`~/\.claude/plans/[^\s]+\.md`)

func Run() {
	cfg, err := config.LoadConfig()
	if err != nil {
		log.Fatalf("Config error: %v", err)
	}

	client := telegram.NewClient(cfg.BotToken)
	botStartTime := time.Now().Unix()
	offset := 0

	log.Println("Bot started, polling for messages...")

	// Signal handling
	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigCh
		log.Println("Shutting down bot...")
		os.Exit(0)
	}()

	for {
		updates, err := client.GetUpdates(offset, 30)
		if err != nil {
			if strings.Contains(err.Error(), "409") {
				log.Fatal("Conflict: another bot instance is running")
			}
			log.Printf("Polling error: %v", err)
			time.Sleep(5 * time.Second)
			continue
		}

		for _, update := range updates {
			offset = update.UpdateID + 1
			if update.Message == nil {
				continue
			}

			msg := update.Message
			chatID := fmt.Sprintf("%d", msg.Chat.ID)

			// Ignore old messages
			if int64(msg.Date) < botStartTime {
				continue
			}

			// Chat ID registration
			if cfg.ChatID == "" {
				cfg.ChatID = chatID
				config.SaveChatID(chatID)
				client.SendMessage(chatID, fmt.Sprintf("Chat ID saved: <code>%s</code>\nYou're all set!", chatID), "HTML")
				log.Printf("Chat ID registered: %s", chatID)
				continue
			}

			// Authorization
			if chatID != cfg.ChatID {
				client.SendMessage(chatID, "Unauthorized.", "")
				continue
			}

			// Photo handling
			if len(msg.Photo) > 0 {
				handlePhoto(client, cfg, chatID, msg)
				continue
			}

			text := strings.TrimSpace(msg.Text)
			if text == "" {
				continue
			}

			// Command routing
			switch {
			case text == "/stop":
				handleStop(client, chatID)
			case text == "/allow":
				handleAllow(client, chatID)
			case text == "/1":
				handleAllow(client, chatID)
			case text == "/deny":
				handleDeny(client, chatID)
			case text == "/escape":
				handleEscape(client, chatID)
			case text == "/status" || text == "/sessions":
				handleStatus(client, chatID)
			case text == "/screen":
				handleScreen(client, chatID)
			case text == "/plan":
				handlePlan(client, chatID)
			case strings.HasPrefix(text, "/switch"):
				handleSwitch(client, chatID, text)
			case strings.HasPrefix(text, "/rename"):
				handleRename(client, chatID, text)
			case text == "/help" || text == "/start":
				handleHelp(client, chatID)
			case optionRegex.MatchString(text):
				handleSelectOption(client, chatID, text)
			default:
				handleInput(client, chatID, text)
			}
		}
	}
}

func handleInput(client *telegram.Client, chatID, text string) {
	active := sessions.GetActive()
	if active == "" {
		client.SendMessage(chatID, "No active session.", "")
		return
	}
	if !tmux.SessionExists(active) {
		client.SendMessage(chatID, fmt.Sprintf("Session <code>%s</code> not found in tmux.", formatter.EscapeHTML(active)), "HTML")
		return
	}
	if err := tmux.SendKeys(active, text, ""); err != nil {
		client.SendMessage(chatID, fmt.Sprintf("Error sending input: %v", err), "")
		return
	}
	client.SendMessage(chatID, fmt.Sprintf("Sent to <b>%s</b>", formatter.EscapeHTML(active)), "HTML")
}

func handlePhoto(client *telegram.Client, cfg *config.Config, chatID string, msg *telegram.Message) {
	active := sessions.GetActive()
	if active == "" {
		client.SendMessage(chatID, "No active session.", "")
		return
	}
	if !tmux.SessionExists(active) {
		client.SendMessage(chatID, "Session not found in tmux.", "")
		return
	}

	// Get largest photo
	photo := msg.Photo[len(msg.Photo)-1]
	file, err := client.GetFile(photo.FileID)
	if err != nil {
		client.SendMessage(chatID, fmt.Sprintf("Failed to get file: %v", err), "")
		return
	}

	uploadDir := filepath.Join(os.Getenv("HOME"), ".claude", "claude-tg", "uploads")
	localPath, err := client.DownloadFile(file.FilePath, uploadDir)
	if err != nil {
		client.SendMessage(chatID, fmt.Sprintf("Failed to download: %v", err), "")
		return
	}

	var text string
	if msg.Caption != "" {
		text = fmt.Sprintf("%s (see image: %s)", msg.Caption, localPath)
	} else {
		text = fmt.Sprintf("Please look at this image: %s", localPath)
	}

	if err := tmux.SendKeys(active, text, ""); err != nil {
		client.SendMessage(chatID, fmt.Sprintf("Error sending: %v", err), "")
		return
	}
	client.SendMessage(chatID, fmt.Sprintf("Photo sent to <b>%s</b>", formatter.EscapeHTML(active)), "HTML")
}

func handleStatus(client *telegram.Client, chatID string) {
	state := sessions.List()
	if len(state.Sessions) == 0 {
		client.SendMessage(chatID, "No active sessions.", "")
		return
	}

	var lines []string
	lines = append(lines, "<b>Sessions:</b>")
	for name := range state.Sessions {
		alive := tmux.SessionExists(name)
		dot := "🔴"
		if alive {
			dot = "🟢"
		}
		if name == state.Active {
			lines = append(lines, fmt.Sprintf("%s <code>%s</code> ◀ active", dot, formatter.EscapeHTML(name)))
		} else {
			lines = append(lines, fmt.Sprintf("%s /switch %s", dot, formatter.EscapeHTML(name)))
		}
	}
	client.SendMessage(chatID, strings.Join(lines, "\n"), "HTML")
}

func handleSwitch(client *telegram.Client, chatID, text string) {
	parts := strings.Fields(text)
	if len(parts) < 2 {
		client.SendMessage(chatID, "Usage: /switch &lt;session-name&gt;", "HTML")
		return
	}
	target := parts[1]
	if err := sessions.SetActive(target); err != nil {
		client.SendMessage(chatID, fmt.Sprintf("Error: %v", err), "")
		return
	}
	client.SendMessage(chatID, fmt.Sprintf("Switched to <b>%s</b>", formatter.EscapeHTML(target)), "HTML")
}

func handleScreen(client *telegram.Client, chatID string) {
	active := sessions.GetActive()
	if active == "" {
		client.SendMessage(chatID, "No active session.", "")
		return
	}
	content, err := tmux.CapturePane(active, 40, "")
	if err != nil {
		client.SendMessage(chatID, fmt.Sprintf("Error: %v", err), "")
		return
	}
	if len(content) > 3800 {
		content = content[:3800] + "..."
	}
	client.SendMessage(chatID, fmt.Sprintf("<pre>%s</pre>", formatter.EscapeHTML(content)), "HTML")
}

func handlePlan(client *telegram.Client, chatID string) {
	active := sessions.GetActive()
	if active == "" {
		client.SendMessage(chatID, "No active session.", "")
		return
	}
	content, err := tmux.CapturePane(active, 50, "")
	if err != nil {
		client.SendMessage(chatID, "Could not capture screen.", "")
		return
	}
	match := planPathRegex.FindString(content)
	if match == "" {
		client.SendMessage(chatID, "No plan file found in terminal output.", "")
		return
	}
	planPath := strings.Replace(match, "~", os.Getenv("HOME"), 1)
	data, err := os.ReadFile(planPath)
	if err != nil {
		client.SendMessage(chatID, fmt.Sprintf("Could not read plan: %v", err), "")
		return
	}
	sendLongMessage(client, chatID, "Plan", string(data))
}

func sendLongMessage(client *telegram.Client, chatID, title, content string) {
	escaped := formatter.EscapeHTML(content)
	full := fmt.Sprintf("<b>%s</b>\n\n<pre>%s</pre>", title, escaped)
	if len(full) <= 4096 {
		client.SendMessage(chatID, full, "HTML")
		return
	}

	// Split into chunks
	lines := strings.Split(escaped, "\n")
	chunk := fmt.Sprintf("<b>%s</b>\n\n<pre>", title)
	first := true
	for _, line := range lines {
		if len(chunk)+len(line)+10 > 3800 {
			chunk += "</pre>"
			client.SendMessage(chatID, chunk, "HTML")
			chunk = "<pre>"
			first = false
		}
		if first && len(chunk) > len(fmt.Sprintf("<b>%s</b>\n\n<pre>", title)) {
			chunk += "\n"
		}
		chunk += line + "\n"
	}
	if strings.TrimSpace(strings.TrimPrefix(chunk, "<pre>")) != "" {
		chunk += "</pre>"
		client.SendMessage(chatID, chunk, "HTML")
	}
}

func handleRename(client *telegram.Client, chatID, text string) {
	parts := strings.Fields(text)
	if len(parts) < 2 {
		client.SendMessage(chatID, "Usage: /rename &lt;new-name&gt;", "HTML")
		return
	}
	newName := parts[1]
	if !nameRegex.MatchString(newName) {
		client.SendMessage(chatID, "Invalid name. Use alphanumeric characters and hyphens.", "")
		return
	}
	active := sessions.GetActive()
	if active == "" {
		client.SendMessage(chatID, "No active session.", "")
		return
	}
	if err := tmux.RenameSession(active, newName); err != nil {
		client.SendMessage(chatID, fmt.Sprintf("tmux rename failed: %v", err), "")
		return
	}
	if err := sessions.Rename(active, newName); err != nil {
		client.SendMessage(chatID, fmt.Sprintf("Session rename failed: %v", err), "")
		return
	}
	client.SendMessage(chatID, fmt.Sprintf("Renamed to <b>%s</b>", formatter.EscapeHTML(newName)), "HTML")
}

func handleStop(client *telegram.Client, chatID string) {
	active := sessions.GetActive()
	if active == "" {
		client.SendMessage(chatID, "No active session.", "")
		return
	}
	tmux.SendInterrupt(active, "")
	client.SendMessage(chatID, fmt.Sprintf("Sent Ctrl+C to <b>%s</b>", formatter.EscapeHTML(active)), "HTML")
}

func handleAllow(client *telegram.Client, chatID string) {
	active := sessions.GetActive()
	if active == "" {
		client.SendMessage(chatID, "No active session.", "")
		return
	}
	tmux.SendEnter(active, "")
	client.SendMessage(chatID, "Approved ✓", "")
}

func handleDeny(client *telegram.Client, chatID string) {
	active := sessions.GetActive()
	if active == "" {
		client.SendMessage(chatID, "No active session.", "")
		return
	}
	tmux.SendEscape(active, "")
	client.SendMessage(chatID, "Denied ✗", "")
}

func handleSelectOption(client *telegram.Client, chatID, text string) {
	matches := optionRegex.FindStringSubmatch(text)
	if matches == nil {
		return
	}
	num, _ := strconv.Atoi(matches[1])
	active := sessions.GetActive()
	if active == "" {
		client.SendMessage(chatID, "No active session.", "")
		return
	}
	if num == 1 {
		handleAllow(client, chatID)
		return
	}
	tmux.SendArrowDown(active, num-1, "")
	time.Sleep(100 * time.Millisecond)
	tmux.SendEnter(active, "")
	client.SendMessage(chatID, fmt.Sprintf("Selected option %d ✓", num), "")
}

func handleEscape(client *telegram.Client, chatID string) {
	active := sessions.GetActive()
	if active == "" {
		client.SendMessage(chatID, "No active session.", "")
		return
	}
	tmux.SendEscape(active, "")
	client.SendMessage(chatID, "Sent Escape", "")
}

func handleHelp(client *telegram.Client, chatID string) {
	help := `<b>Claude TG Commands</b>

<b>Input</b>
Just type — sends text to active session
Send a photo — downloads and passes path to Claude

<b>Responses</b>
/allow — approve permission prompt
/deny — reject permission prompt
/1 /2 /3 — select numbered option
/escape — cancel/dismiss dialog
/stop — send Ctrl+C

<b>Session</b>
/status — list sessions
/switch &lt;name&gt; — switch active session
/rename &lt;name&gt; — rename active session
/screen — capture terminal
/plan — view current plan

/help — this message`
	client.SendMessage(chatID, help, "HTML")
}
